OpenRA.Platforms.Default/OpenAlSoundEngine.cs

```diff
@@ -404,54 +412,256 @@ public void Dispose()
-				return pos / 22050f;
+				int sampleOffset;
+				AL10.alGetSourcei(Source, AL11.AL_SAMPLE_OFFSET, out sampleOffset);
+				return sampleOffset / SampleRate;
 			}
 		}
 
-		public bool Playing
+		public virtual bool Complete
 		{
 			get
 			{
 				int state;
 				AL10.alGetSourcei(Source, AL10.AL_SOURCE_STATE, out state);
-				return state == AL10.AL_PLAYING;
+				return state == AL10.AL_STOPPED;
 			}
 		}
 
 		public void SetPosition(WPos pos)
 		{
 			AL10.alSource3f(Source, AL10.AL_POSITION, pos.X, pos.Y, pos.Z);
 		}
+
+		protected void StopSource()
+		{
+			int state;
+			AL10.alGetSourcei(Source, AL10.AL_SOURCE_STATE, out state);
+			if (state == AL10.AL_PLAYING || state == AL10.AL_PAUSED)
+				AL10.alSourceStop(Source);
+		}
+
+		public virtual void Stop()
+		{
+			StopSource();
+			AL10.alSourcei(Source, AL10.AL_BUFFER, 0);
+		}
+	}
+
+	class OpenAlStreamingSound : OpenAlSound
+	{
+		const int BufferSizeInSecs = 1;
+		readonly object bufferDequeueLock = new object();
+		readonly CancellationTokenSource cts = new CancellationTokenSource();
+		readonly Task streamTask;
+		int totalBuffersPlayed;
+
+		public OpenAlStreamingSound(uint source, bool looping, bool relative, WPos pos, float volume, int channels, int sampleBits, int sampleRate, Stream stream)
+			: base(source, looping, relative, pos, volume, sampleRate)
+		{
+			streamTask = Task.Run(async () =>
+			{
+				var format = OpenAlSoundEngine.MakeALFormat(channels, sampleBits);
+				var bytesPerSample = sampleBits / 8;
+
+				var buffers = new uint[3];
+				AL10.alGenBuffers(new IntPtr(buffers.Length), buffers);
+				try
+				{
+					var freeBuffers = new Stack<uint>(buffers);
+					var data = new byte[sampleRate * bytesPerSample * BufferSizeInSecs];
+					var streamEnd = false;
+
+					while (!streamEnd && !cts.IsCancellationRequested)
+					{
+						// Fill the data array as fully as possible.
+						var count = await ReadFillingBuffer(stream, data);
+						streamEnd = count < data.Length;
+
+						// Fill a buffer and queue it for playback.
+						var nextBuffer = freeBuffers.Pop();
+						AL10.alBufferData(nextBuffer, format, data, new IntPtr(count), new IntPtr(sampleRate));
+						AL10.alSourceQueueBuffers(source, new IntPtr(1), ref nextBuffer);
+
+						lock (cts)
+							if (!cts.IsCancellationRequested)
+							{
+								// Once we have at least one buffer filled, we can actually start.
+								// If streaming fell behind, the source will have stopped,
+								// we also need to play it in this case to resume audio.
+								int state;
+								AL10.alGetSourcei(source, AL10.AL_SOURCE_STATE, out state);
+								if (state == AL10.AL_INITIAL || state == AL10.AL_STOPPED)
+								{
+									// If we resume playback from the stopped state, it resets to the beginning!
+									// To avoid replaying the same audio, we need to dequeue the processed buffers first.
+									if (state == AL10.AL_STOPPED)
+										DequeueBuffers(source, freeBuffers);
```

Comment:
Commenting this out fixes my issues on OSX without introducing any new regressions as far as I can tell (it doesn't reset to the beginning or replay the same audio).